docs/design/phases.md

```diff
@@ -0,0 +1,354 @@
+- `kubeadm phase addons kube-proxy`
+   - Inputs:
+     - `--image=gcr.io/google_containers/kube-proxy-${ARCH}:${STABLE_VERSION}`
+     - `--admin-kubeconfig=/etc/kubernetes/admin.conf`
+     - `--pod-network-cidr=`
+   - Outputs:
+     - Applies the built-in `kube-proxy` manifest to the API Server
+- `kubeadm phase addons kube-dns`
+   - Inputs:
+     - `--admin-kubeconfig=/etc/kubernetes/admin.conf`
+     - `--dns-domain=cluster.local`
+     - `--image-repository=gcr.io/google_containers`
+   - Outputs:
+     - Applies the built-in `kube-dns` manifest to the API Server
+
+### `kubeadm init` UX and function in the light of `kubeadm phase`
+
+ - `kubeadm init`
+   - `--cert-dir=/etc/kubernetes/pki`
+   - `--service-cidr=10.96.0.1/12`
+   - `--cert-altnames=""`
+   - `--apiserver-address=0.0.0.0`
+   - `--apiserver-port=6443`
+   - `--image-repository=gcr.io/google_containers`
+   - `--pod-network-cidr=`
+   - `--dns-domain=cluster.local`
+   - `--token=${AUTOGENERATE}`
+   - `--token-ttl=0`
+   - `--use-kubernetes-version=${STABLE_VERSION}`
+   - `--self-hosted`
+
+What would `kubeadm init` actually do?
+
+```shell
+kubeadm phase certs selfsign \
+	--cert-dir=${CERT_DIR} \
+	--altnames=${CERT_ALTNAMES} \
+	--dns-domain=${DNS_DOMAIN} \
+	--service-cidr=${SERVICE_CIDR}
+
+kubeadm phase kubeconfig clientcert \
+	--client-name=kubernetes-admin \
+	--organization="system:masters" \
+	--api-servers=${APISERVER_ADDRESS}:${APISERVER_PORT} \
+	--cert-dir=${CERT_DIR} \
+	> /etc/kubernetes/admin.conf
+
+kubeadm phase kubeconfig clientcert \
+	--client-name="system:node:$(hostname -i)" \
+	--organization="system:nodes" \
+	--api-servers=${APISERVER_ADDRESS}:${APISERVER_PORT} \
+	--cert-dir=${CERT_DIR} \
+	> /etc/kubernetes/kubelet.conf
+
+kubeadm phase kubeconfig clientcert \
+	--client-name=system:controller-manager \
+	--api-servers=${APISERVER_ADDRESS}:${APISERVER_PORT} \
+	--cert-dir=${CERT_DIR} \
+	> /etc/kubernetes/controller-manager.conf
+
+kubeadm phase kubeconfig clientcert \
+	--client-name=system:kube-scheduler \
+	--api-servers=${APISERVER_ADDRESS}:${APISERVER_PORT} \
+	--cert-dir=${CERT_DIR} \
+	> /etc/kubernetes/scheduler.conf
+
+kubeadm phase controlplane etcd \
+	--image ${IMAGE_REPOSITORY}/etcd-${GOARCH}:3.0.14-kubeadm
+
+kubeadm phase controlplane apiserver \
+	--image=${IMAGE_REPOSITORY}/kube-apiserver-${ARCH}:${USE_KUBERNETES_VERSION} \
+	--service-cidr=${SERVICE_CIDR} \
+	--cert-dir=${CERT_DIR} \
+	--port=${APISERVER_PORT} \
+	--address=${APISERVER_ADDRESS}
+
+kubeadm phase controlplane controller-manager \
+	--image=${IMAGE_REPOSITORY}/kube-controller-manager-${ARCH}:${USE_KUBERNETES_VERSION} \
+	--cluster-cidr=${POD_NETWORK_CIDR} \
+	--cert-dir=${CERT_DIR}
+
+kubeadm phase controlplane scheduler \
+	--image=${IMAGE_REPOSITORY}/kube-scheduler-${ARCH}:${USE_KUBERNETES_VERSION}
+
+kubectl taint node $(hostname -i) node-role.kubernetes.io/master:NoSchedule
+kubectl label node $(hostname -i) node-role.kubernetes.io/master=""
```

Comment:
Using `kubectl`in the middle of a `kubeadm`sequence of `phases`sounds a little bit strange. With low priority (but also low effort), it might make sense to implement a phase also for this step.